### PR TITLE
[fix][fn] enable Go function token auth and TLS

### LINKED ISSUE #20451: [Bug][fn] Go functions runtime does not support any sort of auth.
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Version

After enabling JWT auth on our cluster, all of our Go functions started failing.  This is because authentication information is only passed to the `pulsar-admin` command, and also because the Go function SDK has no concept of auth providers.

### Minimal reproduce step

Just enable auth, and watch your go functions start to complain:
```
broker.pulsar.svc.cluster.local:6650 local_addr=172.27.168.187:37436
2023/06/01 03:05:16.139 connection.go:226: [warning] Failed to establish connection with broker: 'Unable to authenticate'  remote_addr=pulsar://pulsar-broker.pulsar.svc.cluster.local:6650 local_addr=172.27.168.187:37436
2023/06/01 03:05:45.027 log.go:77: [error] create producer error:connection error
2023/06/01 03:05:45.027 instance.go:210: [fatal] connection error
Stream closed EOF for pulsar/pf-public-default-function-test-30-c169f141-0 (pulsarfunction)

```        

### What did you expect to see?

Working functions.

### What did you see instead?

Dead functions.

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by flowchartsman**
Fixes #20451

### Motivation

Go functions die when auth is turned on. I would like them to not die.

### Modifications

Added transport of auth plugin and auth plugin params to go instance config.
Added parsing of these values into the pf sdk

### Verifying this change

- [ ] Make sure that the change passes the CI checks.
- [X] ~Cannot integration test functions with auth~ 

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [X] `doc-not-needed` 
- [ ] `doc-complete` 

No doc change needed, because the docs don't mention the exception, so now the intuitive assumption holds.

### Matching PR in forked repository

PR in forked repository: https://github.com/flowchartsman/pulsar/pull/6

### CODE REVIEW

File: pulsar-functions/runtime/src/test/java/org/apache/pulsar/functions/runtime/RuntimeUtilsTest.java [hunk 1]
```diff
@@ -73,6 +73,9 @@ public void getGoInstanceCmd(boolean k8sRuntime) throws IOException {
         instanceConfig.setPort(1337);
         instanceConfig.setMetricsPort(60000);
 
+        AuthenticationConfig authConfig = AuthenticationConfig.builder().build();
+
+
```
Comment (michaeljmarshall) [inline pulsar-functions/runtime/src/test/java/org/apache/pulsar/functions/runtime/RuntimeUtilsTest.java:78]: Nit: we can remove these two lines.

**Comment by flowchartsman** [comment]
**NOTE** Unless I missed something there are no unit integration tests for functions that enable auth. The closest I could find was the tests in `tests/integration/src/test/java/org/apache/pulsar/tests/integration/functions/PulsarFunctionsTest.java`, however it makes no mention of auth. I'm not familiar enough with the testing framework of the product to know if there's a way to easily run one of these tests with auth enabled, so if we want those tests, I'm going to need some assistance.

**Review by michaeljmarshall — APPROVED** [review]
LGTM. The one detail I see that might be missing for the go function runtime is an ability to configure TLS. Token based auth is recommended to be performed over TLS. 

That feature would go in another PR. At the very least, we'll want to document this limitation to make sure users are aware of the current state of go functions.

**Comment by flowchartsman** [comment]
Converting this to draft to integrate TLS (see #20453 / PIP-273)

**Comment by flowchartsman** [comment]
> That feature would go in another PR.
Hold my 🍺

File: pulsar-function-go/pf/instanceConf.go [hunk 2]
```diff
@@ -42,6 +42,11 @@ type instanceConf struct {
 	killAfterIdle               time.Duration
 	expectedHealthCheckInterval int32
 	metricsPort                 int
+	authPlugin                  string
+	authParams                  string
+	tlsCertsPath                string
```
Comment (michaeljmarshall) [inline pulsar-function-go/pf/instanceConf.go:47]: Nit: I think we want this to be `tlsTrustCertsPath`. In the project, we typically use `tlsCertsPath` to be a server's public certs that get sent to the client.

**Review by michaeljmarshall — APPROVED** [review]
Great work @flowchartsman! I left one comment, but it's very minor.

**Review by michaeljmarshall — APPROVED** [review]

**Review by tisonkun — APPROVED** [review]

**Comment by tisonkun** [comment]
Merging...

Thanks for your contribution @flowchartsman!

**Pull request merged**